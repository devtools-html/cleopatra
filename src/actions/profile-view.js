/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type {
  ProfileSelection, CallTreeFilter, ImplementationFilter,
} from '../types/actions';
import type { Action, ThunkAction } from '../types/store';
import type { ThreadIndex, IndexIntoFuncTable, IndexIntoMarkersTable } from '../types/profile';

/**
 * The actions that pertain to changing the view on the profile, including searching
 * and filtering. Currently the call tree's actions are in this file, but should be
 * split apart. These actions should most likely affect every panel.
 */
export function changeSelectedFuncStack(
  threadIndex: ThreadIndex,
  selectedFuncStack: IndexIntoFuncTable[]
): Action {
  return {
    type: 'CHANGE_SELECTED_FUNC_STACK',
    selectedFuncStack, threadIndex,
  };
}

export function changeSelectedThread(selectedThread: ThreadIndex): Action {
  return {
    type: 'CHANGE_SELECTED_THREAD',
    selectedThread,
  };
}

export function changeThreadOrder(threadOrder: ThreadIndex[]): Action {
  return {
    type: 'CHANGE_THREAD_ORDER',
    threadOrder,
  };
}

export function hideThread(
  threadIndex: ThreadIndex,
  threadOrder: ThreadIndex[],
  hiddenThreads: ThreadIndex[]
): ThunkAction<void> {
  return dispatch => {
    // Do not allow hiding the last thread.
    if (hiddenThreads.length + 1 === threadOrder.length) {
      return;
    }

    dispatch(({
      type: 'HIDE_THREAD',
      threadIndex,
      threadOrder,
      hiddenThreads,
    }: Action));
  };
}

export function showThread(threadIndex: ThreadIndex): Action {
  return {
    type: 'SHOW_THREAD',
    threadIndex,
  };
}

export function changeCallTreeSearchString(searchString: string): Action {
  return {
    type: 'CHANGE_CALL_TREE_SEARCH_STRING',
    searchString,
  };
}

export function changeExpandedFuncStacks(
  threadIndex: ThreadIndex,
  expandedFuncStacks: Array<IndexIntoFuncTable[]>
): Action {
  return {
    type: 'CHANGE_EXPANDED_FUNC_STACKS',
    threadIndex, expandedFuncStacks,
  };
}

export function changeSelectedMarker(
  threadIndex: ThreadIndex,
  selectedMarker: IndexIntoMarkersTable | -1
): Action {
  return {
    type: 'CHANGE_SELECTED_MARKER',
    selectedMarker, threadIndex,
  };
}

export function changeImplementationFilter(implementation: ImplementationFilter): Action {
  return {
    type: 'CHANGE_IMPLEMENTATION_FILTER',
    implementation,
  };
}

export function changeInvertCallstack(invertCallstack: boolean): Action {
  return {
    type: 'CHANGE_INVERT_CALLSTACK',
    invertCallstack,
  };
}

export function changeHidePlatformDetails(hidePlatformDetails: boolean): Action {
  return {
    type: 'CHANGE_HIDE_PLATFORM_DETAILS',
    hidePlatformDetails,
  };
}

export type UpdateProfileSelection = (selection: ProfileSelection) => Action;

export function updateProfileSelection(selection: ProfileSelection): Action {
  return {
    type: 'UPDATE_PROFILE_SELECTION',
    selection,
  };
}

export function addRangeFilter(start: number, end: number): Action {
  return {
    type: 'ADD_RANGE_FILTER',
    start, end,
  };
}

export function addRangeFilterAndUnsetSelection(start: number, end: number): ThunkAction<void> {
  return dispatch => {
    dispatch(addRangeFilter(start, end));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function popRangeFilters(firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_RANGE_FILTERS',
    firstRemovedFilterIndex,
  };
}

export function popRangeFiltersAndUnsetSelection(firstRemovedFilterIndex: number): ThunkAction<void> {
  return dispatch => {
    dispatch(popRangeFilters(firstRemovedFilterIndex));
    dispatch(updateProfileSelection({ hasSelection: false, isModifying: false }));
  };
}

export function addCallTreeFilter(threadIndex: ThreadIndex, filter: CallTreeFilter): Action {
  return {
    type: 'ADD_CALL_TREE_FILTER',
    threadIndex,
    filter,
  };
}

export function popCallTreeFilters(threadIndex: ThreadIndex, firstRemovedFilterIndex: number): Action {
  return {
    type: 'POP_CALL_TREE_FILTERS',
    threadIndex,
    firstRemovedFilterIndex,
  };
}
