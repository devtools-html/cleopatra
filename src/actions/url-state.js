/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action } from '../types/store';

export function urlStateHasChanged(): Action {
  return {
    type: 'URL_STATE_HAS_CHANGED',
  };
}
