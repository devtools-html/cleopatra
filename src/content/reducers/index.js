import profileView from './profile-view';
import app from './app';
import summaryView from './summary-view';
import urlState from './url-state';
import flameChart from './flame-chart';
import timelineView from './timeline-view';

export default { app, profileView, summaryView, urlState, flameChart, timelineView };
