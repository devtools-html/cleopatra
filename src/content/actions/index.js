// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
