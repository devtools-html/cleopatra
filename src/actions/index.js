/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
