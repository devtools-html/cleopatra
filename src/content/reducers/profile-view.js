import { applyFunctionMerging, setFuncNames, setTaskTracerNames } from '../symbolication';
import { combineReducers } from 'redux';
import { createSelector } from 'reselect';
import * as URLState from './url-state';
import * as ProfileData from '../profile-data';
import * as StackTiming from '../stack-timing';
import * as ProfileTree from '../profile-tree';
import * as TaskTracer from '../task-tracer';

function profile(state = {}, action) {
  switch (action.type) {
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_WEB':
    case 'RECEIVE_PROFILE_FROM_FILE':
      return action.profile;
    case 'COALESCED_FUNCTIONS_UPDATE': {
      const { functionsUpdatePerThread } = action;
      const threads = state.threads.map((thread, threadIndex) => {
        if (!functionsUpdatePerThread[threadIndex]) {
          return thread;
        }
        const { oldFuncToNewFuncMap, funcIndices, funcNames } = functionsUpdatePerThread[threadIndex];
        return setFuncNames(applyFunctionMerging(thread, oldFuncToNewFuncMap),
                            funcIndices, funcNames);
      });
      return Object.assign({}, state, { threads });
    }
    case 'ASSIGN_TASK_TRACER_NAMES': {
      const { addressIndices, symbolNames } = action;
      const tasktracer = setTaskTracerNames(state.tasktracer, addressIndices, symbolNames);
      return Object.assign({}, state, { tasktracer });
    }
    default:
      return state;
  }
}

function funcStackAfterCallTreeFilter(funcArray, filter) {
  if (filter.type === 'prefix' && !filter.matchJSOnly) {
    return removePrefixFromFuncArray(filter.prefixFuncs, funcArray);
  }
  return funcArray;
}

function removePrefixFromFuncArray(prefixFuncs, funcArray) {
  if (prefixFuncs.length > funcArray.length ||
      prefixFuncs.some((prefixFunc, i) => prefixFunc !== funcArray[i])) {
    return [];
  }
  return funcArray.slice(prefixFuncs.length - 1);
}

function threadOrder(state = [], action) {
  switch (action.type) {
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_WEB':
    case 'RECEIVE_PROFILE_FROM_FILE':
      return ProfileData.defaultThreadOrder(action.profile.threads);
    case 'CHANGE_THREAD_ORDER':
      return action.threadOrder;
    default:
      return state;
  }
}

function symbolicationStatus(state = 'DONE', action) {
  switch (action.type) {
    case 'START_SYMBOLICATING':
      return 'SYMBOLICATING';
    case 'DONE_SYMBOLICATING':
      return 'DONE';
    default:
      return state;
  }
}

function viewOptionsThreads(state = [], action) {
  switch (action.type) {
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_WEB':
    case 'RECEIVE_PROFILE_FROM_FILE':
      return action.profile.threads.map(() => ({
        selectedFuncStack: [],
        expandedFuncStacks: [],
        selectedMarker: -1,
      }));
    case 'COALESCED_FUNCTIONS_UPDATE': {
      const { functionsUpdatePerThread } = action;
      // For each thread, apply oldFuncToNewFuncMap to that thread's
      // selectedFuncStack and expandedFuncStacks.
      return state.map((thread, threadIndex) => {
        if (!functionsUpdatePerThread[threadIndex]) {
          return thread;
        }
        const { oldFuncToNewFuncMap } = functionsUpdatePerThread[threadIndex];
        const selectedFuncStack = thread.selectedFuncStack.map(oldFunc => {
          const newFunc = oldFuncToNewFuncMap.get(oldFunc);
          return newFunc === undefined ? oldFunc : newFunc;
        });
        const expandedFuncStacks = thread.expandedFuncStacks.map(oldFuncArray => {
          return oldFuncArray.map(oldFunc => {
            const newFunc = oldFuncToNewFuncMap.get(oldFunc);
            return newFunc === undefined ? oldFunc : newFunc;
          });
        });
        return {
          selectedFuncStack,
          expandedFuncStacks,
          selectedMarker: thread.selectedMarker,
        };
      });
    }
    case 'CHANGE_SELECTED_FUNC_STACK': {
      const { selectedFuncStack, threadIndex } = action;
      const expandedFuncStacks = state[threadIndex].expandedFuncStacks.slice();
      for (let i = 1; i < selectedFuncStack.length; i++) {
        expandedFuncStacks.push(selectedFuncStack.slice(0, i));
      }
      return [
        ...state.slice(0, threadIndex),
        Object.assign({}, state[threadIndex], { selectedFuncStack, expandedFuncStacks }),
        ...state.slice(threadIndex + 1),
      ];
    }
    case 'CHANGE_EXPANDED_FUNC_STACKS': {
      const { threadIndex, expandedFuncStacks } = action;
      return [
        ...state.slice(0, threadIndex),
        Object.assign({}, state[threadIndex], { expandedFuncStacks }),
        ...state.slice(threadIndex + 1),
      ];
    }
    case 'CHANGE_SELECTED_MARKER': {
      const { threadIndex, selectedMarker } = action;
      return [
        ...state.slice(0, threadIndex),
        Object.assign({}, state[threadIndex], { selectedMarker }),
        ...state.slice(threadIndex + 1),
      ];
    }
    case 'ADD_CALL_TREE_FILTER': {
      const { threadIndex, filter } = action;
      const expandedFuncStacks = state[threadIndex].expandedFuncStacks.map(fs => funcStackAfterCallTreeFilter(fs, filter));
      const selectedFuncStack = funcStackAfterCallTreeFilter(state[threadIndex].selectedFuncStack, filter);
      return [
        ...state.slice(0, threadIndex),
        Object.assign({}, state[threadIndex], { selectedFuncStack, expandedFuncStacks }),
        ...state.slice(threadIndex + 1),
      ];
    }
    default:
      return state;
  }
}

function waitingForLibs(state = new Set(), action) {
  switch (action.type) {
    case 'REQUESTING_SYMBOL_TABLE': {
      const newState = new Set(state);
      newState.add(action.requestedLib);
      return newState;
    }
    case 'RECEIVED_SYMBOL_TABLE_REPLY': {
      const newState = new Set(state);
      newState.delete(action.requestedLib);
      return newState;
    }
    default:
      return state;
  }
}

function selection(state = { hasSelection: false, isModifying: false }, action) { // TODO: Rename to timeRangeSelection
  switch (action.type) {
    case 'UPDATE_PROFILE_SELECTION':
      return action.selection;
    default:
      return state;
  }
}

function scrollToSelectionGeneration(state = 0, action) {
  switch (action.type) {
    case 'CHANGE_INVERT_CALLSTACK':
    case 'CHANGE_JS_ONLY':
    case 'CHANGE_SELECTED_FUNC_STACK':
    case 'CHANGE_SELECTED_THREAD':
      return state + 1;
    default:
      return state;
  }
}

function rootRange(state = { start: 0, end: 1 }, action) {
  switch (action.type) {
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_WEB':
    case 'RECEIVE_PROFILE_FROM_FILE':
      return ProfileData.getTimeRangeIncludingAllThreads(action.profile);
    default:
      return state;
  }
}

function zeroAt(state = 0, action) {
  switch (action.type) {
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_WEB':
    case 'RECEIVE_PROFILE_FROM_FILE':
      return ProfileData.getTimeRangeIncludingAllThreads(action.profile).start;
    default:
      return state;
  }
}

function tabOrder(state = [0, 1, 2, 3, 4, 5], action) {
  switch (action.type) {
    case 'CHANGE_TAB_ORDER':
      return action.tabOrder;
    default:
      return state;
  }
}

const viewOptions = combineReducers({
  threads: viewOptionsThreads,
  threadOrder, symbolicationStatus, waitingForLibs,
  selection, scrollToSelectionGeneration, rootRange, zeroAt,
  tabOrder,
});

export default combineReducers({ viewOptions, profile });

export const getProfileView = state => state.profileView;

/**
 * Profile View Options
 */
export const getProfileViewOptions = state => getProfileView(state).viewOptions;

export const getScrollToSelectionGeneration = createSelector(
  getProfileViewOptions,
  viewOptions => viewOptions.scrollToSelectionGeneration
);

export const getZeroAt = createSelector(
  getProfileViewOptions,
  viewOptions => viewOptions.zeroAt
);

export const getThreadOrder = createSelector(
  getProfileViewOptions,
  viewOptions => viewOptions.threadOrder
);

export const getDisplayRange = createSelector(
  state => getProfileViewOptions(state).rootRange,
  state => getProfileViewOptions(state).zeroAt,
  URLState.getRangeFilters,
  (rootRange, zeroAt, rangeFilters) => {
    if (rangeFilters.length > 0) {
      let { start, end } = rangeFilters[rangeFilters.length - 1];
      start += zeroAt;
      end += zeroAt;
      return { start, end };
    }
    return rootRange;
  }
);

export const getTasksByThread = createSelector(
  state => getProfileTaskTracerData(state).taskTable,
  state => getProfileTaskTracerData(state).threadTable,
  TaskTracer.getTasksByThread
);

/**
 * Profile
 */
export const getProfile = state => getProfileView(state).profile;
export const getProfileInterval = state => getProfile(state).meta.interval;
export const getThreadNames = state => getProfile(state).threads.map(t => t.name);
export const getProfileTaskTracerData = state => getProfile(state).tasktracer;

const selectorsForThreads = {};

export const selectorsForThread = threadIndex => {
  if (!(threadIndex in selectorsForThreads)) {
    const getThread = state => getProfile(state).threads[threadIndex];
    const getViewOptions = state => getProfileViewOptions(state).threads[threadIndex];
    const getCallTreeFilters = state => URLState.getCallTreeFilters(state, threadIndex);
    const getRangeFilteredThread = createSelector(
      getThread,
      getDisplayRange,
      (thread, range) => {
        const { start, end } = range;
        return ProfileData.filterThreadToRange(thread, start, end);
      }
    );
    const getRangeFilteredThreadSamples = createSelector(
      getRangeFilteredThread,
      thread => thread.samples
    );
    const getRangeFilteredThreadMarkers = createSelector(
      getRangeFilteredThread,
      thread => thread.markers
    );
    const getJankInstances = createSelector(
      getRangeFilteredThreadSamples,
      state => getThread(state).processType,
      (samples, processType) => ProfileData.getJankInstances(samples, processType, 50)
    );
    const getTracingMarkers = createSelector(
      getThread,
      getRangeFilteredThreadMarkers,
      (thread, markers) => ProfileData.getTracingMarkers(thread, markers)
    );
    const getRangeAndCallTreeFilteredThread = createSelector(
      getRangeFilteredThread,
      getCallTreeFilters,
      (thread, callTreeFilters) => {
        const result = callTreeFilters.reduce((t, filter) => {
          switch (filter.type) {
            case 'prefix':
              return ProfileData.filterThreadToPrefixStack(t, filter.prefixFuncs, filter.matchJSOnly);
            case 'postfix':
              return ProfileData.filterThreadToPostfixStack(t, filter.postfixFuncs, filter.matchJSOnly);
            default:
              throw new Error('unhandled call tree filter');
          }
        }, thread);
        return result;
      }
    );
    const getJSOnlyFilteredThread = createSelector(
      getRangeAndCallTreeFilteredThread,
      URLState.getJSOnly,
      (thread, jsOnly) => {
        return jsOnly ? ProfileData.filterThreadToJSOnly(thread) : thread;
      }
    );
    const getJSOnlyAndSearchFilteredThread = createSelector(
      getJSOnlyFilteredThread,
      URLState.getSearchString,
      (thread, searchString) => {
        return ProfileData.filterThreadToSearchString(thread, searchString);
      }
    );
    const getFilteredThread = createSelector(
      getJSOnlyAndSearchFilteredThread,
      URLState.getInvertCallstack,
      (thread, shouldInvertCallstack) => {
        return shouldInvertCallstack ? ProfileData.invertCallstack(thread) : thread;
      }
    );
    const getRangeSelectionFilteredThread = createSelector(
      getFilteredThread,
      getProfileViewOptions,
      (thread, viewOptions) => {
        if (!viewOptions.selection.hasSelection) {
          return thread;
        }
        const { selectionStart, selectionEnd } = viewOptions.selection;
        return ProfileData.filterThreadToRange(thread, selectionStart, selectionEnd);
      }
    );
    const getFuncStackInfo = createSelector(
      getFilteredThread,
      ({stackTable, frameTable, funcTable}) => {
        return ProfileData.getFuncStackInfo(stackTable, frameTable, funcTable);
      }
    );
    const getSelectedFuncStackAsFuncArray = createSelector(
      getViewOptions,
      threadViewOptions => threadViewOptions.selectedFuncStack
    );
    const getSelectedFuncStack = createSelector(
      getFuncStackInfo,
      getSelectedFuncStackAsFuncArray,
      (funcStackInfo, funcArray) => {
        return ProfileData.getFuncStackFromFuncArray(funcArray, funcStackInfo.funcStackTable);
      }
    );
    const getExpandedFuncStacksAsFuncArrays = createSelector(
      getViewOptions,
      threadViewOptions => threadViewOptions.expandedFuncStacks
    );
    const getExpandedFuncStacks = createSelector(
      getFuncStackInfo,
      getExpandedFuncStacksAsFuncArrays,
      (funcStackInfo, funcArrays) => {
        return funcArrays.map(funcArray => ProfileData.getFuncStackFromFuncArray(funcArray, funcStackInfo.funcStackTable));
      }
    );
    const getFuncStackMaxDepth = createSelector(
      getFilteredThread,
      getFuncStackInfo,
      StackTiming.computeFuncStackMaxDepth
    );
    const getStackTimingByDepth = createSelector(
      getFilteredThread,
      getFuncStackInfo,
      getFuncStackMaxDepth,
      getProfileInterval,
      StackTiming.getStackTimingByDepth
    );
    const getCallTree = createSelector(
      getRangeSelectionFilteredThread,
      getProfileInterval,
      getFuncStackInfo,
      URLState.getJSOnly,
      ProfileTree.getCallTree
    );
    selectorsForThreads[threadIndex] = {
      getThread,
      getRangeFilteredThread,
      getViewOptions,
      getCallTreeFilters,
      getFilteredThread,
      getJankInstances,
      getTracingMarkers,
      getRangeSelectionFilteredThread,
      getFuncStackInfo,
      getSelectedFuncStack,
      getExpandedFuncStacks,
      getFuncStackMaxDepth,
      getStackTimingByDepth,
      getCallTree,
    };
  }
  return selectorsForThreads[threadIndex];
};

export const selectedThreadSelectors = (() => {
  const anyThreadSelectors = selectorsForThread(0);
  const result = {};
  for (const key in anyThreadSelectors) {
    result[key] = state => selectorsForThread(URLState.getSelectedThreadIndex(state))[key](state);
  }
  return result;
})();
