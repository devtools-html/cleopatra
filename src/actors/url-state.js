/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import { getUrlState } from '../reducers/url-state';
import {
  urlStateHasChanged,
  updateUrlState,
  show404,
} from '../actions/url-state';
import { urlFromState, stateFromLocation } from '../url-handling';

import type { Dispatch } from '../types/store';
import type { State, UrlState } from '../types/reducers';

let previousUrlState = null;
export default function stateWatcher(state: State, dispatch: Dispatch) {
  function updateState() {
    if (window.history.state) {
      dispatch(updateUrlState(window.history.state));
    } else {
      try {
        const urlState = stateFromLocation(window.location);
        // replacing the existing location ensures we get the latest format for the URL
        updateLocation(urlState, true);
        dispatch(updateUrlState(urlState));
      } catch (e) {
        console.error(e);
        show404(window.location.pathname + window.location.search);
      }
    }
  }

  function initialSetup() {
    updateState();
    window.addEventListener('popstate', updateState);
  }

  function updateLocation(urlState: UrlState, replace: boolean = false) {
    const newUrl = urlFromState(urlState);
    const operation = replace
      ? window.history.replaceState.bind(window.history)
      : window.history.pushState.bind(window.history);
    if (newUrl !== window.location.pathname + window.location.search) {
      operation(urlState, document.title, newUrl);
    }
  }

  const newUrlState = getUrlState(state);
  if (!previousUrlState) {
    // first run
    previousUrlState = newUrlState;
    initialSetup();
  } else if (newUrlState !== previousUrlState) {
    previousUrlState = newUrlState;
    updateLocation(newUrlState);
    dispatch(urlStateHasChanged());
  }
}
