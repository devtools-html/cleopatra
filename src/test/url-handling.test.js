/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
/**
 * @jest-environment jsdom
 */
import * as urlStateReducers from '../reducers/url-state';
import { stateFromLocation, urlStateToUrlObject } from '../url-handling';
import { blankStore } from './fixtures/stores';
import getGeckoProfile from './fixtures/profiles/gecko-profile';
import { processProfile } from '../profile-logic/process-profile';
import { receiveProfileFromStore } from '../actions/receive-profile';
import { selectedThreadSelectors } from '../reducers/profile-view';
import type { Profile } from '../types/profile';
import getProfile from './fixtures/profiles/call-nodes';

function _getStoreWithURL(
  settings: {
    pathname?: string,
    search?: string,
    hash?: string,
  },
  profile: Profile = getProfile()
) {
  const { pathname, hash, search } = Object.assign(
    {
      pathname: '/public/1ecd7a421948995171a4bb483b7bcc8e1868cc57/calltree/',
      hash: '',
      search: '',
    },
    settings
  );
  const urlState = stateFromLocation({
    pathname,
    search,
    hash,
  });
  const store = blankStore();
  store.dispatch({ type: '@@urlenhancer/updateUrlState', urlState });
  store.dispatch(receiveProfileFromStore(profile));
  return store;
}

describe('selectedThread', function() {
  function storeWithThread(threadIndex) {
    const store = blankStore();
    const urlState = stateFromLocation({
      pathname: '/public/1ecd7a421948995171a4bb483b7bcc8e1868cc57/calltree/',
      search: `?thread=${threadIndex}`,
      hash: '',
    });
    store.dispatch({ type: '@@urlenhancer/updateUrlState', urlState });

    return store;
  }

  it('selects the right thread when receiving a profile from web', function() {
    const profile: Profile = processProfile(getGeckoProfile());

    const store = storeWithThread(1);
    store.dispatch(receiveProfileFromStore(profile));

    expect(urlStateReducers.getSelectedThreadIndex(store.getState())).toBe(1);
  });

  it('selects a default thread when a wrong thread has been requested', function() {
    const profile: Profile = processProfile(getGeckoProfile());

    const store = storeWithThread(100);
    store.dispatch(receiveProfileFromStore(profile));

    // "2" is the content process' main tab
    expect(urlStateReducers.getSelectedThreadIndex(store.getState())).toBe(2);
  });
});

describe('threadOrder and hiddenThreads', function() {
  const profileWithThreads: Profile = processProfile(getGeckoProfile());

  it('decides the threadOrder and hiddenThreads without any set parameters', function() {
    const { getState } = _getStoreWithURL({ search: '' }, profileWithThreads);
    expect(urlStateReducers.getThreadOrder(getState())).toEqual([0, 2, 1]);
    expect(urlStateReducers.getHiddenThreads(getState())).toEqual([]);
  });

  it('can reorder the threads ', function() {
    const { getState } = _getStoreWithURL(
      { search: '?threadOrder=1-2-0' },
      profileWithThreads
    );
    expect(urlStateReducers.getThreadOrder(getState())).toEqual([1, 2, 0]);
    expect(urlStateReducers.getHiddenThreads(getState())).toEqual([]);
  });

  it('can hide the threads', function() {
    const { getState } = _getStoreWithURL(
      { search: '?hiddenThreads=1-2' },
      profileWithThreads
    );
    expect(urlStateReducers.getThreadOrder(getState())).toEqual([0, 2, 1]);
    expect(urlStateReducers.getHiddenThreads(getState())).toEqual([1, 2]);
  });

  it('will not accept invalid threads in the thread order', function() {
    const { getState } = _getStoreWithURL(
      { search: '?threadOrder=0-8-2-a-1' },
      profileWithThreads
    );
    expect(urlStateReducers.getThreadOrder(getState())).toEqual([0, 2, 1]);
    expect(urlStateReducers.getHiddenThreads(getState())).toEqual([]);
  });

  it('will not accept invalid hidden threads', function() {
    const { getState } = _getStoreWithURL(
      { search: '?hiddenThreads=0-8-2-a' },
      profileWithThreads
    );
    expect(urlStateReducers.getThreadOrder(getState())).toEqual([0, 2, 1]);
    expect(urlStateReducers.getHiddenThreads(getState())).toEqual([0, 2]);
  });
});

describe('url upgrading', function() {
  /**
   * Originally transform stacks were called call tree filters. This test asserts that
   * the upgrade process works correctly.
   */
  describe('version 1: legacy URL serialized call tree filters', function() {
    it('can upgrade callTreeFilters to transforms', function() {
      const { getState } = _getStoreWithURL({
        search:
          '?callTreeFilters=prefix-012~prefixjs-123~postfix-234~postfixjs-345',
      });
      const transforms = selectedThreadSelectors.getTransformStack(getState());
      expect(transforms).toEqual([
        {
          type: 'focus-subtree',
          callNodePath: [0, 1, 2],
          implementation: 'combined',
          inverted: false,
        },
        {
          type: 'focus-subtree',
          callNodePath: [1, 2, 3],
          implementation: 'js',
          inverted: false,
        },
        {
          type: 'focus-subtree',
          callNodePath: [2, 3, 4],
          implementation: 'combined',
          inverted: true,
        },
        {
          type: 'focus-subtree',
          callNodePath: [3, 4, 5],
          implementation: 'js',
          inverted: true,
        },
      ]);
    });
  });

  describe('version 2: split apart timeline tab', function() {
    it('switches to the flame chart when given a timeline tab', function() {
      const { getState } = _getStoreWithURL({
        pathname: '/public/e71ce9584da34298627fb66ac7f2f245ba5edbf5/timeline/',
      });
      expect(urlStateReducers.getSelectedTab(getState())).toBe('flame-chart');
    });

    it('switches to the markers-table when given a markers tab', function() {
      const { getState } = _getStoreWithURL({
        pathname: '/public/e71ce9584da34298627fb66ac7f2f245ba5edbf5/markers/',
      });
      expect(urlStateReducers.getSelectedTab(getState())).toBe('markers-table');
    });
  });
});

describe('URL serialization of the transform stack', function() {
  const transformString =
    'f-combined-012~ms-js-123~mcn-combined-234~f-js-345-i~mf-6~ff-7~cr-combined-8-9';
  const { getState } = _getStoreWithURL({
    search: '?v=1&transforms=' + transformString,
  });

  it('deserializes focus subtree transforms', function() {
    const transformStack = selectedThreadSelectors.getTransformStack(
      getState()
    );

    // The indexes don't necessarily map to anything logical in the profile fixture.
    expect(transformStack).toEqual([
      {
        type: 'focus-subtree',
        callNodePath: [0, 1, 2],
        implementation: 'combined',
        inverted: false,
      },
      {
        type: 'merge-subtree',
        callNodePath: [1, 2, 3],
        implementation: 'js',
        inverted: false,
      },
      {
        type: 'merge-call-node',
        callNodePath: [2, 3, 4],
        implementation: 'combined',
      },
      {
        type: 'focus-subtree',
        callNodePath: [3, 4, 5],
        implementation: 'js',
        inverted: true,
      },
      {
        type: 'merge-function',
        funcIndex: 6,
      },
      {
        type: 'focus-function',
        funcIndex: 7,
      },
      {
        type: 'collapse-resource',
        resourceIndex: 8,
        collapsedFuncIndex: 9,
        implementation: 'combined',
      },
    ]);
  });

  it('re-serializes the focus subtree transforms', function() {
    const { query } = urlStateToUrlObject(
      urlStateReducers.getUrlState(getState())
    );
    expect(query.transforms).toBe(transformString);
  });
});
