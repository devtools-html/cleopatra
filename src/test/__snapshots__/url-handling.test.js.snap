// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`symbolServerUrl will error when switching to an invalid host 1`] = `
Array [
  Array [
    "The symbol server URL was not valid, defaulting to https://symbols.mozilla.org/symbolicate/v5",
    [TypeError: Invalid URL: invalid],
  ],
]
`;

exports[`symbolServerUrl will error when switching to an unknown host 1`] = `
Array [
  Array [
    "The symbol server URL was not in the list of allowed domains, defaulting to https://symbols.mozilla.org/symbolicate/v5",
  ],
]
`;
